<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>AsmVar by ShujiaHuang</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <script src="javascripts/scale.fix.js"></script>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1 class="header">AsmVar</h1>
        <p class="header"></p>

        <ul>
          <li class="download"><a class="buttons" href="https://github.com/ShujiaHuang/AsmVar/zipball/master">Download ZIP</a></li>
          <li class="download"><a class="buttons" href="https://github.com/ShujiaHuang/AsmVar/tarball/master">Download TAR</a></li>
          <li><a class="buttons github" href="https://github.com/ShujiaHuang/AsmVar">View On GitHub</a></li>
        </ul>

        <p class="header">This project is maintained by <a class="header name" href="https://github.com/ShujiaHuang">ShujiaHuang</a></p>


      </header>
      <section>
        <h1>
<a id="asmvar" class="anchor" href="#asmvar" aria-hidden="true"><span class="octicon octicon-link"></span></a>AsmVar</h1>

<p>AsmVar is a software for discovery, genotyping and characterization of structural variants and novel sequence at single nucleotide resolution from de novo assemblies on a population scale</p>

<p><strong>Contributors:</strong> Shujia Huang, Siyang Liu, Junhua Rao &amp; Weijian Ye (Contribute equally)  <br>
<strong>Contact              :</strong> <a href="mailto:liusiyang@genomics.cn">liusiyang@genomics.cn</a> &amp; <a href="mailto:huangshujia@genomics.cn">huangshujia@genomics.cn</a>                <br>
<strong>Institute            :</strong> BGI &amp; KU in GenomeDK consortium                 <br>
<strong>Last Version         :</strong> 2014-10-30                                      <br></p>

<h2>
<a id="usage" class="anchor" href="#usage" aria-hidden="true"><span class="octicon octicon-link"></span></a>Usage</h2>

<ol>
<li>Overview</li>
<li>How to use AsmVar</li>
<li>Speed and Memory required</li>
<li>Install the software</li>
<li>Run the software</li>
<li>Software output </li>
<li>Please cite the paper: AsmVar: Discovery, genotyping and characterization of structural variants and novel sequence at single nucleotide resolution from de novo assemblies on a population scale (manuscript in preparation)</li>
<li>Acknowledgement : Bioinformatics teams in GenomeDK consortium</li>
</ol>

<h2>
<a id="instant-plans-for-further-developments----" class="anchor" href="#instant-plans-for-further-developments----" aria-hidden="true"><span class="octicon octicon-link"></span></a>Instant plans for further developments    </h2>

<ol>
<li>Streamline evaluations of de novo assemblies with regards to continuity, completeness and accuracy.<br>
</li>
<li>Assign quality scores to de novo assemblies<br>
</li>
<li>Improve genotyping module and develop new approaches for greatest accuracy and flexibility.<br>
</li>
<li>Streamline the characterization of structural variants modules in AsmVar<br>
</li>
<li>Improvement of de novo assemblies<br>
</li>
</ol>

<h2>
<a id="contribution" class="anchor" href="#contribution" aria-hidden="true"><span class="octicon octicon-link"></span></a>Contribution</h2>

<p>The AsmVar is initially developed for the DanishPanGenome project in the GenomeDK platform. The AsmVar framework construction, applications, statistical methods and the evaluation protocols are established by <a href="https://github.com/SiyangLiu">Siyang Liu</a> and <a href="https://github.com/ShujiaHuang">Shujia Huang</a>, Junhua Rao and <a href="https://github.com/WeijianYe">Weijian Ye</a> led by <a href="https://github.com/SiyangLiu">Siyang Liu</a> and <a href="https://github.com/ShujiaHuang">Shujia Huang</a>.  The coding contributions for different modules are written in the source code title. Most of the initial codes are written in C++, python and perl by <a href="https://github.com/ShujiaHuang">Shujia Huang</a> and perl, python and R by <a href="https://github.com/SiyangLiu">Siyang Liu</a>(All modules especially SV discovery and genotyping), Junhua Rao (especially SV mechanism), <a href="https://github.com/WeijianYe">Weijian Ye</a> (especially Ancestral state and Novel sequence) . <a href="https://github.com/ShujiaHuang">Shujia Huang</a> and <a href="https://github.com/SiyangLiu">Siyang Liu</a> take charge of the software architecture and the efficiency of the algorithms. The work is supervised by <a href="http://www.binf.ku.dk/staff/?pure=en/persons/8330">Anders Krogh</a>, Jun Wang and Xun Xu.</p>

<h2>
<a id="license" class="anchor" href="#license" aria-hidden="true"><span class="octicon octicon-link"></span></a>LICENSE</h2>

<p>Released under the <a href="http://opensource.org/licenses/MIT">MIT License</a>
Copyright Â© 2014-2015</p>
      </section>
      <footer>
        <p><small>Hosted on <a href="http://pages.github.com">GitHub Pages</a> using the Dinky theme</small></p>
      </footer>
    </div>
    <!--[if !IE]><script>fixScale(document);</script><![endif]-->
		
  </body>
</html>
