<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8' />
    <meta http-equiv="X-UA-Compatible" content="chrome=1" />
    <meta name="description" content="AsmVar : " />

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>AsmVar</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/ShujiaHuang/AsmVar">View on GitHub</a>

          <h1 id="project_title">AsmVar</h1>
          <h2 id="project_tagline"></h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/ShujiaHuang/AsmVar/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/ShujiaHuang/AsmVar/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h1>
<a name="asmvar" class="anchor" href="#asmvar"><span class="octicon octicon-link"></span></a>AsmVar</h1>

<p>AsmVar is a software for discovery, genotyping and characterization of structural variants and novel sequence at single nucleotide resolution from de novo assemblies on a population scale</p>

<p><strong>Author &amp; contributors:</strong> Shujia Huang, Siyang Liu, Junhua Rao &amp; Weijian Ye (Contribute equally)  <br><strong>Contact              :</strong> <a href="mailto:liusiyang@genomics.cn">liusiyang@genomics.cn</a>                           <br><strong>Institute            :</strong> BGI &amp; KU in GenomeDK consortium                 <br><strong>Last Version         :</strong> 2014-05-30                                      <br></p>

<h2>
<a name="usage" class="anchor" href="#usage"><span class="octicon octicon-link"></span></a>Usage</h2>

<ol>
<li>Overview</li>
<li>How to use AsmVar</li>
<li>Speed and Memory required</li>
<li>Install the software</li>
<li>Run the software</li>
<li>Software output </li>
<li>Please cite the paper: AsmVar: Discovery, genotyping and characterization of structural variants and novel sequence at single nucleotide resolution from de novo assemblies on a population scale (manuscript in preparation)</li>
<li>Acknowledgement : Bioinformatics teams in GenomeDK consortium</li>
</ol><h2>
<a name="instant-plans-for-further-developments" class="anchor" href="#instant-plans-for-further-developments"><span class="octicon octicon-link"></span></a>Instant plans for further developments</h2>

<ol>
<li>Streamline evaluations of de novo assemblies with regards to continuity, completeness and accuracy.</li>
<li>Assign quality scores to de novo assemblies</li>
<li>Improve genotyping module and develop new approaches for greatest accuracy and flexibility.</li>
<li>Streamline the characterization of structural variants modules in AsmVar</li>
<li>Improvement of de novo assemblies </li>
</ol><h2>
<a name="contribution" class="anchor" href="#contribution"><span class="octicon octicon-link"></span></a>Contribution</h2>

<p>The AsmVar is developed for the DanishPanGenome project in the GenomeDK platform. The AsmVar framework construction, applications, statistical methods and the evaluation protocols are established by <a href="https://github.com/SiyangLiu">Siyang Liu</a> and <a href="https://github.com/ShujiaHuang">Shujia Huang</a>, Junhua Rao and <a href="https://github.com/WeijianYe">Weijian Ye</a> led by <a href="https://github.com/SiyangLiu">Siyang Liu</a> and <a href="https://github.com/ShujiaHuang">Shujia Huang</a>.  The coding contributions for different modules are written in the source code title. Most of the initial codes are written in C++, python and perl by <a href="https://github.com/ShujiaHuang">Shujia Huang</a> and perl, python and R by <a href="https://github.com/SiyangLiu">Siyang Liu</a>(All modules especially SV discovery and genotyping), Junhua Rao (especially SV mechanism), <a href="https://github.com/WeijianYe">Weijian Ye</a> (especially Ancestral state and Novel sequence) . <a href="https://github.com/ShujiaHuang">Shujia Huang</a> and <a href="https://github.com/SiyangLiu">Siyang Liu</a> take charge of the software architecture and the efficiency of the algorithms. The AsmVar is under active development at the moment, dedicated to provide population-scale exact breakpoint variants called from de novo assemblies. This work is supervised by <a href="http://www.binf.ku.dk/staff/?pure=en/persons/8330">Anders Krogh</a>, Xun Xu, Jun Wang with critical input and suggestions from Mikkel Schierup, Simon Rasmussen, Palle V Fredsted, and Soren Besenbacher and José MG Izarzugaz.</p>

<h2>
<a name="license" class="anchor" href="#license"><span class="octicon octicon-link"></span></a>LICENSE</h2>

<p>Released under the <a href="http://www.opensource.org/licenses/MIT">MIT License</a>
Copyright <g-emoji alias="copyright" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/00a9.png">©</g-emoji> 2014-2015</p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">AsmVar maintained by <a href="https://github.com/ShujiaHuang">ShujiaHuang</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
