Asmvar
==========

Asmvar is a software for detecting variants, including Indel & SV, base on long sequence(or de novo Assembly) alignement.

LICENSE 
--------
Copyright &copy; 2014-2015

__Author & contributors:__ Shujia Huang, Siyang Liu, Weijian Ye & Junhua Rao   <br/>
__Contact              :__ liusiyang.ocean@gmail.com & huangshujia@genomics.cn <br/>
__Institute            :__ BGI & KU                                            <br/>
__Last Version         :__ 2014-05-30                                          <br/>

Contribution
------------

The AsmVar is developed for the DanishPanGenome project in GenomeDK platform. This manuscript is written by Siyang Liu independently supervised by Anders Krogh.  The AsmVar software construction, applications, statistical methods and the evaluation protocols are established by Siyang Liu and Shujia Huang and with discussions from Junhua Rao and Weijian Ye.  The coding contributions for different modules are written in the source code title. Most of the initial codes are written in C++, python, perl by Shujia Huang and perl, python and R by Siyang Liu, Weijian Ye and Junhua Rao. Shujia Huang takes charge of the software architecture and greatly improves the efficiency of algorithms. The AsmVar is under active development at the moment, dedicated to provide population-scale exact breakpoint variants called from de novo assemblies or long read technology such as illumine molecule and CG LFR. 

Usage
-----
1. Overview
2. How to use Asmvar
3. Speed and Memory required
4. Install the software
5. Run the software
6. Software output 
7. Please cite the paper ..., if you need .
8. Acknowledgement :

